### DISCUSSION TITLE
Can a package link JavaScript code (using --js-library)?

### DISCUSSION THREAD

**Discussion by pthom** [pyodide/pyodide#5088] [Q&A] [answered]
Hello,

Based on what I observed, JavaScript code that is added via the compilation flag [--js-library](https://emscripten.org/docs/tools_reference/emcc.html#emcc-js-library) is not included in generated wheels, when trying to build a package.

Is there a possibility to include such code at link time?

----
**More details below**

This is a follow up to [this issue](https://github.com/pyodide/pyodide/issues/4724), where I tried and failed to create a WebGL2 context when using SDL. The reason I failed is deep inside SDL (it initializes a bad OpenGL context), and I did not find a solution.

I'm now exploring an alternative, where I would replace SDL by glfw. 

_I tried two ways to do it:_

### 1. Using GLFW provided within emscripten
I first tried using the standard GLFW provided by emscripten, via -sUSE_GLFW=3:

In this case I get an error in the browser console: 
```
Uncaught (in promise) Error: Dynamic linking error: cannot resolve symbol glfwSetErrorCallback
```

(I tried to add `-lglfw3` to cflags and ldflags in meta.yaml, but it did not help).

### 2. Using a new port of glfw: [emscripten-glfw](https://github.com/pongasoft/emscripten-glfw). 
This one is more promising, since it is more recent. However I still get dynamic linking errors
```
Uncaught (in promise) Error: Dynamic linking error: cannot resolve symbol emscripten_glfw3_context_get_now
```

The principle of this library is that the Glfw API is reproduced in javascript: see [lib_emscripten_glfw3.js](https://github.com/pongasoft/emscripten-glfw/blob/master/src/js/lib_emscripten_glfw3.js)
On top of this, there is a small c++ library that will call these: see its [source](https://github.com/pongasoft/emscripten-glfw/tree/master/src/cpp)

The missing function "emscripten_glfw3_context_get_now" is _implemented in javascript,_ not in C++.

I can compile and link succesfully with the C++ library, by using emscripten-glfw's [CMakeLists](https://github.com/pongasoft/emscripten-glfw/blob/master/CMakeLists.txt).

However, in the build log I can see that a flag `--js-library` is added:

```
[12/12] : && /tmp/tmpwoebx35a/c++ -fPIC -O3 -DNDEBUG  -fexceptions -flto -sMAX_WEBGL_VERSION=2 -sMIN_WEBGL_VERSION=2 -lGL --js-library /home/pascal/dvp/_Bundle/_Pyodide/pyodide/packages/daft2/build/daft2-0.0.1/external/emscripten-glfw/src/js/lib_emscripten_glfw3.js -shared  -o _daft2.cpython-312-x86_64-linux-gnu.so CMakeFiles/_daft2.dir/src/python_bindings/module.cpp.o  libdaft2.a  external/emscripten-glfw/libglfw3.a && cd /tmp/tmpitklc4im/build && /tmp/tmpwoebx35a/strip /tmp/tmpitklc4im/build/_daft2.cpython-312-x86_64-linux-gnu.so
```

This flag should add the js code to the output at link time.

However when I unzip the wheel, the javascript is not present

```
.
├── daft2
│   ├── _daft2.cpython-312-wasm32-emscripten.so
│   └── __init__.py
├── daft2-0.0.1-cp312-cp312-pyodide_2024_0_wasm32.whl
└── daft2-0.0.1.dist-info
    ├── METADATA
    ├── RECORD
    └── WHEEL
  ```

I guess this goes deep inside the build machinery, and that it is not easy to add, but I would be interested in an answer if someone knows more about this.

Thanks!

**Comment by ryanking13** [comment, marked as answer]
Emscripten has a Main module and Side module concept. The main module consists of WASM modules plus JavaScript codes, while side modules are pure WASM modules. During the runtime, a main module is loaded first, and it can load side modules dynamically on demand.

In Pyodide case, the main module is pyodide.asm.[js, wasm] and packages are side modules. So you cannot link JS libraries into the package and it needs to be linked to the main module.

We are linking JS libraries [here](https://github.com/pyodide/pyodide/blob/3222f2ed2d57bd34a8b55276c0cb333997843c42/Makefile.envs#L143-L181), so I think you can start linking JavaScript part of GLFW there (`-lglfw.js`).

Note that linking JS libraries to the main module is not enough and you'll also need to link the binary parts of the GLFW to your package. (using `-sUSE_GLFW=3` might work).

**Comment by pthom** [comment]
Magnificent! 
Thanks a lot, your answers are always precise and to the point. This is precious!

**Reply by hoodmane** [reply]
It is possible to inject some js code into side modules which will be `eval`ed when the dynamic library is loaded. If this sounds useful I can mock up an example.

**Comment by pthom** [comment]
@hoodmane : thank you! At the moment, it is good enough. 

By the way, I switched back to SDL (which is good since it is more maintained): I found the 2 reasons for my previous failures with SDL:
- In order to link with SDL2, I had to pass `-L$(WASM_LIBRARY_DIR)/lib` [inside meta.yaml](https://github.com/pyodide/pyodide/discussions/4707#discussioncomment-10783478)
- In order to create a context on the C++ side you shall not create it before on the JavaScript side. I'm sorry I had missed this 

As a consequence, I closed a [related issue](https://github.com/pyodide/pyodide/issues/4724) I had opened here.
